### ISSUE TITLE
Auto-toggling Bluetooth

### ISSUE THREAD

**Issue by Avamander** [Freeyourgadget/Gadgetbridge#519]
Would it be possible for GadgetBridge to automatically disable Bluetooth when connection is lost for a while and then toggle it once in a while to see if the device has reappeared? (This would require root most likely)

**Comment by danielegobbetti** [comment]
This would probably require a profound change in the codebase (you might have noticed that gadgetbridge turn its background service off when you toggle Bluetooth) but I'm leaving this open while waiting for confirmations.

**Comment by IzzySoft** [comment]
Raw idea: Even currently, GB tries to reconnect in intervals, starting with small time frames and continually increasing the wait until reaching a maximum. There could be an option that, when reaching the maximum, switching BT off for an interval defined in the configuration – and "backgrounding" instead of "turning off the service" (some marker in the database could be used to differentiate between those states).

Again, after having tried and failed for "max", GB could revert to the "old logic" and consider BT off for good (might require another setting or not).

Side note: I don't see this making sense with too small intervals – unless the purpose behind this is something else than saving battery. Minimal interval probably should be something around 5 min then, if not more. But then, it would (as a side-effect) solve the issue of my Pebble not automatically reconnecting in the morning until I toggle BT off and on again, some days :stuck_out_tongue_winking_eye:

**Comment by cpfeiffer** [comment]
What would be the purpose of this? Would that be a workaround for connection problems or would that be a batter saving mechanism?

**Comment by IzzySoft** [comment]
@cpfeiffer Exactly my question. But to answer it: catching two birds with one stone, I'd say it could be both. For me the former, as I think the latter is rather negligible.

**Comment by Avamander** [comment]
My phone uses about 2x as much battery when Bluetooth is enabled, when my Pebble is not in range I would find it rather optimal that GadgetBridge disabled it.

**Comment by cpfeiffer** [comment]
Frankly, I don't like the idea very much. Randomly toggling Bluetooth just to check if a device is available can have so many side effects (other apps doing something when suddenly Bluetooth is enabled). I'd like to keep that a conscious user action.

Besides, as Daniele already mentioned, this would complicate the whole thing quite a bit. 

I've never heard of a device consuming twice as much energy when Bluetooth is enabled. If that is indeed that case, I assume that either there is a bug in your ROM or there is another app or service doing something all the time when Bluetooth is on.

The only case I see is the connection problem, but maybe that can be fixed differently, either
- by actually fixing the problem, or
- by using something like Tasker to enable Bluetooth in the morning

**Comment by IzzySoft** [comment]
I agree with you, Carsten, and prefer the first bullet-point :stuck_out_tongue_winking_eye:

The second one is a bit tricky – as on what condition would you bind it? If you're a robot getting up each morning at the exact same time (on the second), that'd be easy. But as soon as there's a deviation of ~1h, it's not as easy as sying "toggle BT twice at 7am". And I can tell, as I've tried that. At the moment, I have *Tasker* switching BT off at night and on at 7am. For the Pebble, that seems to work every other day, but not reliably. Despite of always being in BT reach (though at its borders).

If, OTOH, GB would provide Tasker support, so I could have a condition like "8am && BT on && no device connected to GB", that would give me some leverage to play with :stuck_out_tongue:

**Comment by cpfeiffer** [comment]
Do you have a log of such a "morning-connection-problem"?

If you'd use Alarms to wake you up, you might be able to couple that with Tasker (I actually haven't used it myself). If we could add support for wake up alarms to the Pebble, we might use that to toggle BT on the mobile. But I really prefer option one :-)

**Comment by IzzySoft** [comment]
@cpfeiffer I can send you today's log. BT got switched on at 7am, watch still is not connected at 11:50. First appearance of some GB entry in the logs is at 10 am:

    I/nodomain.freeyourgadget.gadgetbridge.externalevents.TimeChangeReceiver( 9949): Time or Timezone changed, syncing with device

Shortly followed by WeatherNotificationReceiver, and then, at 10:05 the first mention of a Pebble being seen:

    I/nodomain.freeyourgadget.gadgetbridge.service.devices.pebble.PebbleIoThread( 9949): found service UUID {HEX}

Repeated 4 times. Pebble still not connected until I manually trigger BT on the Android device again. Though I've noticed on some days it suddenly was found a few hours later.

I can send you the full log per mail, don't want to post it publically.

**Comment by Avamander** [comment]
Though one function in related to this would be auto-reenabling BT when Airplane mode is swiched on (of course via a dialog box but still).
